sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala

```diff
@@ -64,4 +66,10 @@ object DataSourceUtils {
 
   private[sql] def isDataFile(fileName: String) =
     !(fileName.startsWith("_") || fileName.startsWith("."))
+
+  def needRebaseDateTime(lookupFileMeta: String => String): Option[Boolean] = {
+    // If there is no version, we return None and let the caller side to decide.
+    // Files written by Spark 3.0 and later follow the new calendar and don't need to rebase.
+    Option(lookupFileMeta(SPARK_VERSION_METADATA_KEY)).map(_ < "3.0")
```

Comment:
Thank you so much, @cloud-fan . This is much better solution.

Reply:
> Files written by Spark 3.0 and later follow the new calendar and don't need to rebase.

This is an unexpected solution for me which doesn't cover some use case from my point of view. I do believe versions in parquet/avro are not enough to make decision about rebasing.

Reply:
I think the use case you pointed out (write with 3.0 and read with 2.4) is a bit artificial. If someone develops a data pipeline, it's very tricky if some parts of the pipeline use 3.0 and some parts use 2.4. IMO the more common problem is to deal with legacy data files.

BTW we can support this use case by creating a new metadata entry to record the calendar, but I think it's too complicated and not worth it.

Reply:
I don't think we guarantee forward compatibility - things from Spark 2.4 should work in Spark 3.0 but I don't think the opposite case is guaranteed.